### PR TITLE
Prevents saving anonymous field from donation as null and makes anonymous from recurring donations false by default and not nullable

### PR DESCRIPTION
**Pull request by clarissalimab**

### CODE REVIEW

File: lib/insert/insert_donation.rb [hunk 1]
```diff
@@ -173,7 +173,7 @@ def self.insert_payment(kind, fee_total, donation_id, data)
   def self.insert_donation(data, entities)
     d = Donation.new
     d.date = data['date']
-    d.anonymous = data['anonymous']
+    d.anonymous = data['anonymous'].nil? ? false : true
```
Comment (clarissalimab) [inline lib/insert/insert_donation.rb]: We could add the `:before_save` I did before callback instead, I'm not sure which one would be better.

**Comment by clarissalimab** [comment]
After merging this we're going to need to run the script to replace null anonymous again!

File: lib/insert/insert_donation.rb (hunk 1 above)
Comment (wwahammy) [reply lib/insert/insert_donation.rb]: `:before_save` would be better. I'm surprised it doesn't work already though. But I guess I misremembered it. Thanks for catching this!

Comment (clarissalimab) [reply lib/insert/insert_donation.rb]: Oops, I just realized this is not going to work, I'm going to fix it

Comment (clarissalimab) [reply lib/insert/insert_donation.rb]: Okay, I'm going to change for `:before_save` then!

File: lib/insert/insert_donation.rb [hunk 2]
```diff
@@ -173,7 +173,7 @@ def self.insert_payment(kind, fee_total, donation_id, data)
   def self.insert_donation(data, entities)
     d = Donation.new
     d.date = data['date']
-    d.anonymous = data['anonymous']
+    d.anonymous = data['anonymous'].nil? ? false : data['anonymous']
```
Comment (wwahammy) [inline lib/insert/insert_donation.rb]: I think it'd be more logical if it was in a `:before_save`.

You might consider doing : `data['anonymous'] || false` But I'm not sure that's actually as clear.

File: app/models/donation.rb [hunk 3]
```diff
@@ -48,4 +50,8 @@ class Donation < ActiveRecord::Base
 	belongs_to :event
 
 	scope :anonymous, -> {where(anonymous: true)}
+
+	private def set_anonymous
```
Comment (wwahammy) [inline app/models/donation.rb]: Minor nitpick but best practices for ruby generally have the private on it's own line. That, of course, means everything in the file after it is private.

**Review by wwahammy — CHANGES_REQUESTED** [review]
Good work, just a minor editing tweak.

After that's done, if the tests pass, let's merge and deploy.

File: lib/insert/insert_recurring_donation.rb [hunk 4]
```diff
@@ -196,6 +196,7 @@ def self.convert_donation_to_recurring_donation(donation_id)
  def self.insert_recurring_donation(data, entities)
     rd = RecurringDonation.new
     rd.amount = data[:amount]
+    rd.anonymous = data[:anonymous] || false
```
Comment (wwahammy) [inline lib/insert/insert_recurring_donation.rb]: I think this should be an `:after_save` callback too.

**Review by wwahammy — CHANGES_REQUESTED** [review]
Nice!

One little fix I think and also please fix the conflict with `supporter_level_goal`

File: app/models/recurring_donation.rb [hunk 5]
```diff
@@ -15,7 +17,8 @@ class RecurringDonation < ActiveRecord::Base
     :cancelled_at, # datetime of user/supporter who made the cancellation
     :donation_id, :donation,
     :nonprofit_id, :nonprofit,
-    :supporter_id #used because things are messed up in the datamodel
+    :supporter_id, #used because things are messed up in the datamodel
+    :anonymous
```
Comment (wwahammy) [inline app/models/recurring_donation.rb]: Was there a particular reason you needed to add the `:anonymous`?

**Review by wwahammy — CHANGES_REQUESTED** [review]
So, this makes me wonder about something: how is `recurring_donation.anonymous` actually used?

File: app/models/recurring_donation.rb (hunk 5 above)
Comment (clarissalimab) [reply app/models/recurring_donation.rb]: I was getting a `Can't mass-assign protected attributes`

Comment (wwahammy) [reply app/models/recurring_donation.rb]: Ah, alright.

**Comment by clarissalimab** [comment]
I got the impression that it's not being used but I was afraid to break something I was not aware of.

**Comment by wwahammy** [comment]
Could you put an issue in tix for us to evaluate if it can be removed?

**Comment by wwahammy** [comment]
It looks good.

**Pull request merged**